### PR TITLE
[hackathon] additional utilities

### PR DESCRIPTION
**Pull request by beanow-at-crabnebula**
Depends on #240

## Showing attributes

Log entries and spans, now show their attributes.
image
image

## Span / trace error status
Inspired by how opentelemetry [supports a status](https://opentelemetry.io/docs/specs/otel/trace/api/#set-status).

Spans can now have an error status.
(When that span has *any* "log" event with `level = error`)
image

Alternatively they can be orange in the list view, if one of the child spans has an error status.
image

## Detecting "kind" or known trace types ahead of time

Currently DevTools detects a set of known trace types commonly seen in Tauri apps.
It does this reactively, triggered by some of the rendering paths.

Here I've changed this to be done whenever new spans are received for a particular trace.
That trace (or root span) is marked "dirty" and detection will be run again. Avoiding a hot loop of detection by doing it per update batch. And not needing to run it for traces that have already settled and got no new data.

See `clients/web/src/lib/span/update-spans.ts` for this.

## Span details are now tree-shaped

Previously traces were shown with a list of spans.
Now they are:
- Displayed in a tree structure.
- With siblings sorted by start time.
- Colored based on error status & level (dark = debug, very dark = trace).

image

## Associated logs

Logs that have a parent span will be matched against spans in this trace.
So the trace details will now show all related logs.

image

## Jump to source

When a log entry has a source reference part of your own application (e.g. not logged by a 3rd party crate).
You can jump directly to the source file by clicking on the location.

image

## Jump to trace / move to parent

On the console view, you can now directly go to the span details, if it was part of a span.

image

And from a particular span, you can navigate up to it's root span.

image

## Filtering spans

Spans can now by filtered by text. (Case sensitive though.)

image

Or by it's error status (currently includes both direct errors and child span errors).

image

## Parsing plugin commands (new!)

Plugins have a special kind of `cmd` argument like `plugin:your_plugin|some_cmd`.
We now parse those to display in a less confusing way.

image

Given that Tauri v2 is almost entirely based on plugins, this should be particularly helpful there.

Closes DT-153

### CODE REVIEW

File: NEW_STUFF.md
Comment (CrabNejonas) [inline NEW_STUFF.md:1]: Remove?

Comment (johann-crabnebula) [reply NEW_STUFF.md:1]: Removed

File: DEVELOPING.md [hunk 1]
```diff
@@ -10,34 +10,44 @@ There are a couple components to devtools that you should know about:
 
 ## The `cargo-deny` check is failing! What do I do?
 
-The [`cargo-deny`](https://github.com/EmbarkStudios/cargo-deny) action checks for various dependency-related issues, such as problematic licenses, CVEs and the like. Sometimes when opening a PR this check just fails. It's usually a good idea to inspect its output to see exactyl _why_ the check is failing.
+The [`cargo-deny`](https://github.com/EmbarkStudios/cargo-deny) action checks for various dependency-related issues,
+such as problematic licenses, CVEs and the like. Sometimes when opening a PR this check just fails. It's usually a good
+idea to inspect its output to see exactyl _why_ the check is failing.
```
[the remark this answers is not in the document]
Comment (johann-crabnebula) [reply DEVELOPING.md:15]: As of my understanding that is exactly what is explained below.

File: clients/web/src/components/calls/detail-pane/args.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+          {" "}
+          <For each={values()}>
+            {(arg) => {
+              return (
+                <For
+                  each={Object.entries(
+                    typeof arg === "string" ? JSON.parse(arg) : arg,
+                  )}
+                >
+                  {([k, v]) => (
```
[the remark this answers is not in the document]
Comment (johann-crabnebula) [reply clients/web/src/components/calls/detail-pane/args.tsx]: This is some code I only moved but I do agree that key and value is more readable. Changed it.

File: clients/web/src/components/sources/code-view.tsx [hunk 3]
```diff
@@ -1,13 +1,13 @@
-import { Suspense, createResource } from "solid-js";
+import { JSXElement, Suspense, createEffect, createResource } from "solid-js";
```
[the remark this answers is not in the document]
Comment (johann-crabnebula) [reply clients/web/src/components/sources/code-view.tsx]: Took your suggestion. Thanks!

File: clients/web/src/components/icons/migrate--alt.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
+export default function MigrateAlt() {
+  return (
```
[the remark this answers is not in the document]
Comment (johann-crabnebula) [reply clients/web/src/components/icons/migrate--alt.tsx:2]: Added a speaking <title> tag to make the content of the svg more clear.

File: clients/web/src/components/calls/detail-pane/traces.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  spanChildren: SpanTreeEntry[];
+}) {
+  const closedSpans = () =>
+    props.spanChildren.filter((s) => s.span.closedAt > 0);
+
+  const durations = () => {
+    return {
+      start: Math.min(...closedSpans().map((s) => s.span.createdAt)),
+      end: Math.max(...closedSpans().map((s) => s.span.closedAt)),
+      shortest: Math.min(...closedSpans().map((s) => s.span.duration)),
```
[the remark this answers is not in the document]
Comment (johann-crabnebula) [reply clients/web/src/components/calls/detail-pane/traces.tsx]: Good catch. I fixed the regression.

File: clients/web/src/components/calls/list/table-header.tsx [hunk 6]
```diff
@@ -43,12 +43,6 @@ export function CallsListTableHeader(props: {
     });
   };
 
-  const isCurrentlySorted = (column: Column) => {
```
[the remark this answers is not in the document]
Comment (johann-crabnebula) [reply clients/web/src/components/calls/list/table-header.tsx]: Good catch. I fixed the regression.

**Review by alex-crabnebula — APPROVED** [review]

File: clients/web/src/components/calls/detail-pane/detail.tsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  const children = createMemo(() => {
+    const level = minLevel();
+    return spanTreeEntries(
+      props.call,
+      (span) => span.metadata && span.metadata.level <= level,
+    );
+  });
+
+  const includedSpans = createMemo(
+    () => new Set(children().map((n) => n.span.id)),
```
[the remark this answers is not in the document]
Comment (alex-crabnebula) [reply clients/web/src/components/calls/detail-pane/detail.tsx:28]: Let's optimize the associatedLogs instead of the includedSpans.

**Review by alex-crabnebula — APPROVED** [review]

**Pull request merged**